homeassistant/components/viam/services.py

```diff
@@ -0,0 +1,282 @@
+    }
+)
+
+
+def __get_manager(hass: HomeAssistant, call: ServiceCall) -> ViamManager:
+    entry_id: str = call.data[ATTR_CONFIG_ENTRY]
+    entry: ConfigEntry | None = hass.config_entries.async_get_entry(entry_id)
+
+    if not entry:
+        raise ServiceValidationError(
+            f"Invalid config entry: {entry_id}",
+            translation_domain=DOMAIN,
+            translation_key="invalid_config_entry",
+            translation_placeholders={
+                "config_entry": entry_id,
+            },
+        )
+    if entry.state != ConfigEntryState.LOADED:
+        raise ServiceValidationError(
+            f"{entry.title} is not loaded",
+            translation_domain=DOMAIN,
+            translation_key="unloaded_config_entry",
+            translation_placeholders={
+                "config_entry": entry.title,
+            },
+        )
+
+    manager: ViamManager = hass.data[DOMAIN][entry_id]
+    return manager
+
+
+async def __capture_data(call: ServiceCall, *, hass: HomeAssistant) -> None:
+    """Accept input from service call to send to Viam."""
+    manager: ViamManager = __get_manager(hass, call)
+    parts: list[RobotPart] = await manager.get_robot_parts()
+    values = [call.data.get(SERVICE_VALUES)]
+    component_type = call.data.get(SERVICE_COMPONENT_TYPE, "sensor")
+    component_name = call.data.get(SERVICE_COMPONENT_NAME)
+
+    await manager.viam.data_client.tabular_data_capture_upload(
+        tabular_data=values,
+        part_id=parts.pop().id,
+        component_type=component_type,
+        component_name=component_name,
+        method_name="capture_data",
+        data_request_times=[(datetime.now(), datetime.now())],
+    )
+
+
+async def __capture_image(call: ServiceCall, *, hass: HomeAssistant) -> None:
+    """Accept input from service call to send to Viam."""
+    manager: ViamManager = __get_manager(hass, call)
+    parts: list[RobotPart] = await manager.get_robot_parts()
+    filepath = call.data.get(SERVICE_FILEPATH)
+    camera_entity = call.data.get(SERVICE_CAMERA)
+    component_name = call.data.get(SERVICE_COMPONENT_NAME)
+    file_name = call.data.get(SERVICE_FILE_NAME, "camera")
+
+    if filepath is not None:
+        await manager.viam.data_client.file_upload_from_path(
+            filepath=filepath,
+            part_id=parts.pop().id,
+            component_name=component_name,
+        )
+    if camera_entity is not None:
+        image = await camera.async_get_image(hass, camera_entity)
+        await manager.viam.data_client.file_upload(
+            part_id=parts.pop().id,
+            component_name=component_name,
+            file_name=file_name,
+            file_extension=".jpeg",
+            data=image.content,
+        )
+
+
+async def __get_classifications(
+    call: ServiceCall, *, hass: HomeAssistant
+) -> ServiceResponse:
+    """Accept input configuration to request classifications."""
+    manager: ViamManager = __get_manager(hass, call)
```

Comment:
`__get_detections` and this function have a lot of common code. Can we create a common function for it and remove duplicated code?